worker/uniter/uniter.go

```diff
@@ -306,6 +309,15 @@ func (u *Uniter) loop(unitTag names.UnitTag) (err error) {
 		return err
 	}
 
+	// Check we are running the correct charm version.
+	if u.embedded && u.enforcedCharmModifiedVersion != -1 {
+		if charmModifiedVersion != u.enforcedCharmModifiedVersion {
+			u.logger.Infof("remote charm modified version (%d) does not match agent's (%d)",
```

Comment:
I'm unclear on when this could happen.

Reply:
Pod starts up, user upgrades, unit agent starts with old CharmModifiedVersion but the Application in state is different. Uniter downloads the new charm when it should be running the old charm.